markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,321 @@
+Node, this value is not set.
+
+### Tracking It Down
+
+FastBoot works by turning an Ember application and all of its
+dependencies (including your addon) into a single bundle of JavaScript.
+Similar to how you ship an `app.js` and `vendor.js` file to the browser
+when you deploy an Ember app, the FastBoot server also loads these two
+files.
+
+What this means in practice is that errors will include stack traces
+with very large line numbers. Your addon's code will be intermingled
+with other addons, and Ember itself, in the `vendor.js` file.
+
+To figure out what is causing the exception to be thrown, look at the
+stack trace shown when running the app. Note the line number, then open
+`dist/fastboot/vendor.js` in your text editor. Scroll down to the
+provided line number (or, better, use your editor's "Go to Line"
+functionality) to pinpoint exactly which line is causing the issue.
+
+![Example error message and stack trace from an incompatible addon](/images/addon-author-guide/stack-trace-example.png)
+
+### How to Fix It
+
+There are two common approaches to fixing code that accesses browser
+APIs.
+
+First, you can check for the existence of the API you need before
+accessing it. For example, imagine we have an addon that checks the user
+agent of the browser to detect the platform. That code looks something
+like this:
+
+```js
+var ua = window.navigator.userAgent || '';
+```
+
+This line will throw an exception, because `window.navigator` doesn't
+exist in Node. Instead, we can add an extra guard to verify it exists,
+and default to an empty string if not:
+
+```js
+var ua = window && window.navigator ? window.navigator.userAgent : '';
+```
+
+The second option is to move any code that relies on browser-only APIs to
+hooks that only get invoked in the browser.
+
+Independent from FastBoot compatibility, it's generally a good idea to
+do as little work as possible during the evaluation of your addon's
+JavaScript files. Doing upfront work can have a negative impact on load
+times. If you can defer doing work (such as feature detection) until the
+last possible moment, you can improve the performance of applications
+using your addon. Adding FastBoot compatibility is a good opportunity to
+make these improvements to your addon.
+
+Here's a concrete example. Imagine our addon has a component that needs
+to listen to `resize` events on the window. We set up a single listener in
+our component file that's shared across all instances of the component:
+
+```js
+// addon/components/resizable-component.js
+import Component from "ember-component";
+import $ from "jquery";
+
+let componentsToNotify = [];
+$(window).on('resize', () => {
+  componentsToNotify.forEach(c => c.windowDidResize());
+});
+
+export default Component.extend({
+  init() {
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    componentsToNotify.removeObject(this);
```

Comment:
Should avoid prototype extension methods given what we know about the issues with fastboot